csi/pkg/storage/modelregistry_provider.go

```diff
@@ -110,19 +105,84 @@ func (p *ModelRegistryProvider) DownloadModel(modelDir string, modelName string,
 	return provider.DownloadModel(modelDir, "", *modelArtifact.Uri)
 }
 
-func extractProtocol(storageURI string) (kserve.Protocol, error) {
+// Possible URIs:
+// (1) model-registry://{modelName}
+// (2) model-registry://{modelName}/{modelVersion}
+// (3) model-registry://{modelRegistryUrl}/{modelName}
```

Comment:
I am not completely sure this is actually supported, because if you have just 2 tokens (after the trim) it will return the default apiClient (i.e., using the default env model registry).

I think here the problem would be, how do you know whether the user is providing option 2 or 3?

I tried running `make SOURCE_URI=model-registry://model-registry-url/model DEST_PATH=./ run` and in fact `model-registry-url` is interpreted as being the registered model, which is actually not what the user was trying to do.

I think that we have two options here:
1. If the `model-registry-url` is provided, users cannot omit the `version`
2. We use a different delimiter for the `model-registry-url`, e.g., `model-registry://{modelRegistryUrl}:{modelName}`

Reply:
The way I approached this problem can be found here:

https://github.com/Al-Pragliola/model-registry/blob/feat/multi-mr-registries-csi-support/csi/pkg/modelregistry/api_client.go#L24

We try to reach (token 2) as a model registry and on failure we assume that it is a model name and not a valid mr url, it's not perfect but I think it might be a fair compromise

Reply:
Sorry missed that, I tried and I can confirm that if the "host" is reachable it works as expected:
```bash
$ make SOURCE_URI=model-registry://localhost:8080/mymodel DEST_PATH=./ run                                                                                             [17:59:56]
"/usr/bin/go" fmt ./...
"/usr/bin/go" vet ./...
"/usr/bin/go" run ./main.go model-registry://localhost:8080/mymodel ./
2024/10/24 18:00:02 Initializing, args: src_uri [model-registry://localhost:8080/mymodel] dest_path[ [./]
2024/10/24 18:00:02 Download model indexed in model registry: modelName=, storageUri=model-registry://localhost:8080/mymodel, modelDir=./
2024/10/24 18:00:02 Fetching model: registeredModelName=mymodel, versionName=<nil>
2024/10/24 18:00:02 404 Not Found
exit status 1
make: *** [Makefile:59: run] Error 1
```

With a model registry running at `localhost:8080`.

My main concern with this assumption is that, if there is a model registry running but for any reason it is not accessible/reachable we are going to make the wrong assumption that it is a registeredModel name , right? And the error would be misleading to the user as it will find in the logs.

What do you think?

Reply:
Yep that's true, the only hint that user have from the logs is this:

```golang
log.Printf("Falling back to base url %s for model registry service", cfg.Host)
```
We can improve this message telling the user that it failed to reach <url from uri> and it's going to use <fallback url>

---

I wanted this to be as retrocompatible as possible, otherwise there are many alternatives, like a query parameter approach:

`model-registry://url?modelName=x&modelVersion=y`
`model-registry://?modelName=x&modelVersion=y`

or the other delimiter you mentioned.

Reply:
Added comment below for line 129 my2c

Reply:
btw forgot to mention; options 3-4 also consistent with standard URLs and also KServe examples

```
    - regex: "https://(.+?).blob.core.windows.net/(.+)"
    - regex: "https://(.+?).file.core.windows.net/(.+)"
```

much appreciated @Al-Pragliola